programs/drift/src/math/orders.rs

```diff
@@ -825,37 +825,41 @@ pub fn calculate_max_perp_order_size(
         );
     }
 
-    let mut order_size = free_collateral
-        .safe_sub(OPEN_ORDER_MARGIN_REQUIREMENT.cast()?)?
-        .safe_mul(BASE_PRECISION_I128 / QUOTE_PRECISION_I128)?
-        .safe_mul(MARGIN_PRECISION_U128.cast()?)?
-        .safe_div(margin_ratio.cast()?)?
-        .safe_mul(PRICE_PRECISION_I128)?
-        .safe_div(oracle_price_data_price.cast()?)?
-        .safe_mul(PRICE_PRECISION_I128)?
-        .safe_div(quote_oracle_price.cast()?)?
-        .cast::<u64>()?;
-
-    let updated_margin_ratio = perp_market
-        .get_margin_ratio(
-            worst_case_base_asset_amount
-                .unsigned_abs()
-                .safe_add(order_size.cast()?)?,
-            MarginRequirementType::Initial,
-        )?
-        .max(user_custom_margin_ratio);
-
-    if updated_margin_ratio != margin_ratio {
-        order_size = free_collateral
+    let calculate_order_size_and_margin_ratio = |margin_ratio: u32| {
+        let new_order_size = free_collateral
             .safe_sub(OPEN_ORDER_MARGIN_REQUIREMENT.cast()?)?
             .safe_mul(BASE_PRECISION_I128 / QUOTE_PRECISION_I128)?
             .safe_mul(MARGIN_PRECISION_U128.cast()?)?
-            .safe_div(updated_margin_ratio.cast()?)?
+            .safe_div(margin_ratio.cast()?)?
             .safe_mul(PRICE_PRECISION_I128)?
             .safe_div(oracle_price_data_price.cast()?)?
             .safe_mul(PRICE_PRECISION_I128)?
             .safe_div(quote_oracle_price.cast()?)?
             .cast::<u64>()?;
+
+        let new_margin_ratio = perp_market
+            .get_margin_ratio(
+                worst_case_base_asset_amount
+                    .unsigned_abs()
+                    .safe_add(new_order_size.cast()?)?,
+                MarginRequirementType::Initial,
+            )?
+            .max(user_custom_margin_ratio);
+
+        Ok((new_order_size, new_margin_ratio))
+    };
+
+    let mut order_size = 0_u64;
+    let mut updated_margin_ratio = margin_ratio;
+    for _ in 0..6 {
```

Comment:
using iter method like in sdk? jw, is this going to always under estimate the max (to ensure it goes through?)